### ISSUE TITLE
reportOptionalOperand=false does not accept optionals as the RHS of the expression

### ISSUE THREAD

**Issue by Bruntaz** [microsoft/pyright#6687]
**Describe the bug**
When running pyright (version `1.1.339`) with the `reportOptionalOperand=false` flag, I would expect to be able to pass an optional type to either side of a binary operator and have the type error be ignored. The documentation is as follows:

> reportOptionalOperand [boolean or string, optional]: Generate or suppress diagnostics for an attempt to use an Optional type as an operand to a binary or unary operator (like +, ==, or, not). The default value for this setting is "error".

The actual behaviour is that passing an optional to the left hand side has the error silenced, but passing to the right hand side still shows the error (but the error is `reportGeneralTypeIssues`, not `reportOptionalOperand`).

**Code or Screenshots**
[Link to a playground showing the behaviour](https://pyright-play.net/?code=MQAgDgngTglg5gCwC4C4RQKZgPZSQeTCRmwDsBDAG0IynNIBMQBeEAMyoGcMAoHgY0rlOnEAEEUPENJAMMbEAH1FMUjCTKAFN0psAlJJlGQOtgDoARixABGKcemmz-NKqQgAPiAByZDNd9SXntpOQUGbQxdAxDjGAUnK2ZWJxdYh2kzLPSZeJMo8ysAanzdZ0MMoyyzHOk8xJAAPhSC8trjavb61qtG0vM0yqrsyu6yqwAeFrLBocyRjLHCkAn%2Btrn5mtiwkAxI6IqHJecWacLDoc7RhNb%2BEBLEi8qrxZuZprPLJ4yXo7eBpprCzfBy-OL-E5TIEgjoLP5rO6rR7tYY1IA)

```python
# pyright: reportOptionalOperand = false

class A:
    def __init__(self):
        self.b = 1
        self.c: int | None = None

    def d(self):
        if self.b == self.c:  # No error
            ...
        if self.b + self.c:  # Error
            ...
        if self.b >= self.c:  # Error
            ...
        if self.b > self.c:  # Error
            ...
        if self.b <= self.c:  # Error
            ...
        if self.b < self.c:  # Error
            ...

    def e(self):
        if self.c == self.b:  # No error
            ...
        if self.c + self.b:  # No error
            ...
        if self.c >= self.b:  # No error
            ...
        if self.c > self.b:  # No error
            ...
        if self.c <= self.b:  # No error
            ...
        if self.c < self.b:  # No error
            ...
```

Would it be possible to allow this behaviour? If not, can the documentation be updated to specify it only suppresses for the LHS of the operator?

**Comment by erictraut** [comment]
This is the intended behavior, so I don't consider this a bug. I've updated the documentation to clarify that this applies only to the LHS operand.

When using a binary operator in Python, the LHS type determines which class (or classes, in the case of a union type) are used for looking up the operator method (e.g. `__add__` or `__eq__`). The allowed RHS operand is then determined by the annotated signature of the corresponding magic method. Often, these magic methods have multiple (overloaded) methods.

**Comment by Bruntaz** [comment]
> I've updated the documentation to clarify that this applies only to the LHS operand.

Thanks very much for doing that. 

Conceptually though, I'm not sure I agree with how this flag _should_ work. I understand that Python chooses the implementation of the binary method based on the LHS, but it ultimately results in the same error regardless of the order of the operands:

```python
3 < None  # TypeError: '<' not supported between instances of 'int' and 'NoneType'
None < 3  # TypeError: '<' not supported between instances of 'NoneType' and 'int'

class A:
    def __lt__(self, other):
        if isinstance(other, A):
            return True
        else:
            return NotImplemented
A() < None  # TypeError: '<' not supported between instances of 'A' and 'NoneType'
None < A()  # TypeError: '<' not supported between instances of 'NoneType' and 'A'
```

It is equally unsafe to insert `None` on either side of a binary operator that doesn't accept it. On binary operators where it _is_ accepted, they should be properly type annotated to accept the optional type, so `reportOptionalOperand=false` would not be needed.

I feel like the aim of the `reportOptionalXXX` flags should be to tell Pyright to assume correctness on all uses of optionals (i.e. "I promise I can guarantee it won't be `None` when this is actually run"). They should absolutely default to `true`, but if this case is treated as `reportGeneralTypeIssues` then the error can't be silenced without turning off a lot more real type issues.
